### DISCUSSION TITLE
Publication of Cineast Modules on Maven Central

### DISCUSSION THREAD

**Discussion by sauterl** [vitrivr/cineast#177] [Ideas]
In order to prepare module releases on maven central, I suggest some modifications on the current module system:

## Cineast Core

`cineast-core` is _the_ core module: It contains internal representations and an API for the the extraction and retrieval pipelines. Utilities such as extraction / retrieval configuration management (and representations) as well as an API for database binding is also included.

Custom extractors / retrievers require this module as a dependency.

### Proposed Changes

- Move everything that is not part of the above specified _core_ functionality / pipeline APIs, in particular configuration of other elements.
- Potentially rename to `cineast-commons`

## Runtime

`cineast-runtime` is the module that actually has the extraction and retrieval pipelines and runtimes thereof. Also providing the CLI for basic, manual interaction. Depending on [`cineast-core`](#Cineast%20Core).

In an embedded usage, custom projects epend on this module.

### Proposed Changes

- Move runtime related configurations to this module
- Remove custom importers for various dedicated projects to separate project

## External API

`cineast-api` is the module providing the external APIs; WebSocket, REST and gRPC and their endpoints. It depends on [`cineast-runtime`](#Runtime)

Basically _the_ Cineast module for most purposes

### Proposed Changes

- Move API related configurations to this module
- Provide build pipeline for OpenAPI Specifications to be automatically published

---

There are two key points to discuss here:

1. What is preventing us from publishing the modules
2. The minor preparatory steps in order to publish

**Comment by lucaro** [comment]
Publishing these modules to maven central has been on the todo list for a while already. I'm not sure if there's actually anything really preventing us from doing this already or if just nobody got around to actually do it. I personally prefer 'core' over 'commons' since the module contains the primary components and 'commons' suggests more of a collection of utilities to me.
